### ISSUE TITLE
Blockbase: Ponyfill css overrides block-level global styles for Button block

### ISSUE THREAD

**Issue by jeyip** [Automattic/themes#4435]
## Description

I'm seeing an issue with block-level global styling being overridden in the dotcom site editor and frontend (atomic + simple sites), and it looks like it has to do with the [ponyfill.css](https://github.com/Automattic/themes/blob/trunk/blockbase/assets/ponyfill.css) file in `blockbase`. When I remove `ponyfill.css`, per block global styles work as expected.

**List of affected blocks**
1. Button (text color and background color)
2. ~Cover (text color and background color)~ [Not caused by blockbase](https://github.com/Automattic/themes/issues/4435#issuecomment-901524216)
3. ~Site Title (text color and background color)~ [Not caused by blockbase](https://github.com/Automattic/themes/issues/4435#issuecomment-902577643)
4. Cover and Group blocks (right/left padding)

## Steps to replicate

1. On a simple or atomic site, activate a block theme that depends on `blockbase`
2. Activate the site editor
3. Navigate to the site editor
4. Insert a button block onto the page and add text content
5. Open the global styles sidebar, and select the "By Block Type" menu tab
6. Unfold the Button block-level global styles panel and select different color options
7. Confirm that the color options do not change according to user customizations

### Button Block
![2021-08-16 13 58 39](https://user-images.githubusercontent.com/5414230/129790854-598db288-59b6-4a7b-abba-1c24df7e48d7.gif)

## Result
User customized global styles (like background color) weren't applied to the block

## Expected
The block styles would change according to user customizations

## Notes
* Slack Discussion: p1629226103080900-slack-C029FM1EH

**Comment by pbking** [comment]
The button portion of the related ponyfill is [here](https://github.com/Automattic/themes/blob/trunk/blockbase/sass/blocks/_button.scss).  That relates to the video shown but it's not the only place where this is happening.

This is happening because many of Blockbase's selectors have a higher specificity than the styles generated by Global Styles.  This is especially obvious with the button which has a rich ponyfill system including hover states (which are going to be particularly problematic to address in this situation).

**Comment by jffng** [comment]
I believe this affects the editor and the view slightly differently. 

In the case of the button (ignoring the hover states issue), we could theoretically solve the issue by remove the wrapping selector `.wp-block-button` and the global styles customizations win because they load later in the cascade. @mikachan and I tested this and it works in the editor. 

However, in the view, the theme CSS is loaded _after_ the global styles CSS, so the theme CSS is still winning. This is a mismatch between editor and view, and also not the preferred order. We opened an issue above to see if we can fix this upstream.

Front-end source: 

<img width="885" alt="Screen Shot 2021-08-18 at 10 50 54 AM" src="https://user-images.githubusercontent.com/5375500/129920479-6d0eef3d-f363-4dd8-913a-54842dc63dea.png">

**Comment by mikachan** [comment]
I had a quick look at this in Gutenberg and then in Blockbase. I couldn't change the order of `global-styles-inline-css` in Gutenberg other than moving it above `wp-block-library-css`. I did this in [this file](https://github.com/WordPress/gutenberg/blob/trunk/lib/global-styles.php#L337) by changing the priority like so:

```
add_action( 'wp_enqueue_scripts', 'gutenberg_experimental_global_styles_enqueue_assets', 999 );
```

I did manage to force the order via Blockbase, by again changing the priority of the styles in [functions.php](https://github.com/Automattic/themes/blob/trunk/blockbase/functions.php#L67):

```
add_action( 'wp_enqueue_scripts', 'blockbase_scripts', 0 );
```

Which did this:

![Screenshot 2021-08-18 at 16 56 19](https://user-images.githubusercontent.com/1645628/129931336-cf7146db-e623-4be5-b7e4-6f64658a01e3.png)

But I guess this isn't ideal as every theme would need to cater for this.

Thought I'd report back in case this helps in any way, or makes it quicker for someone to work some Gutenberg magic.

**Comment by creativecoder** [comment]
Thanks for investigating @mikachan !

It looks like [WP 5.8 is registering the 'global-styles' stylesheet](https://github.com/wordpress/wordpress-develop/blob/5.8/src/wp-includes/script-loader.php#L2321) as well, which is taking precedence over Gutenberg's enqueue calls. But Gutenberg should be able to unregister and reregister the asset to work around this. I'll spin up a Gutenberg PR.

**Comment by creativecoder** [comment]
Potential fix for this at https://github.com/WordPress/gutenberg/pull/34147. As pointed out earlier, Blockbase styles will still need to be updated to have a css selector specificity equal or below Global Styles css.

**Comment by jeyip** [comment]
It looks like issues with global styles and the cover block can be attributed to specificity problems in core Gutenberg, and not from themes. https://github.com/WordPress/gutenberg/issues/34150. I'll be taking a closer look tomorrow.

**Comment by jeyip** [comment]
Issues with global styles and the site title block are caused by CSS specificity conflicts with `seedlet-blocks`, not `blockbase`.

- Block level global styles are being applied to the site title through `.wp-block-site-title a`
- The Seedlet blocks theme has a competing declaration via `.is-root-container a:not(.wp-block-button__link):not([download])`

https://github.com/Automattic/themes/blob/00ee734b4b0a0001e5f03ad0a431cf032b8436f1/seedlet-blocks/assets/theme.css#L148-L157

I drafted a [separate issue](https://github.com/Automattic/themes/issues/4447) and a [separate PR](https://github.com/Automattic/themes/pull/4446) to address this.

**Comment by creativecoder** [comment]
I found an additional conflict with Blockbase and block specific Global Styles, right/left padding on the Group and Cover blocks are being overridden by Blockbase.

(Global Styles css on the bottom, Blockbase css on the top in both images)

**Cover block**

<img width="437" alt="Screen Shot 2021-08-21 at 06 18 47" src="https://user-images.githubusercontent.com/1699996/130320234-329f55df-9829-432b-af2b-8e18bde32154.png">

**Group block**

<img width="439" alt="Screen Shot 2021-08-21 at 06 19 18" src="https://user-images.githubusercontent.com/1699996/130320244-cf6eaf0b-4bfd-461d-8956-b685d36de07f.png">

**Comment by pbking** [comment]
Thanks @creativecoder.  That (slightly separate) issue is being dealt with in either [this PR ](https://github.com/Automattic/themes/pull/4448)or maybe [this one](https://github.com/Automattic/themes/pull/4454).

**Comment by pbking** [comment]
I think that everything Button-related that CAN be solved has been via #4439 (with the known exceptions of hover state and outline variation).  

Blockbase alignment issues are being looked at this week in #4456

Going to close this one as complete.  Please reopen if anything not tracked elsewhere continues to be an issue.
